src/main/java/org/openrefine/extensions/commons/importer/RelatedCategoryFetcher.java

```diff
@@ -0,0 +1,120 @@
+
+        String titles = fileRecordOriginal.get(0).fileName;
+        int titlesIndex =1;
+        if (titlesIndex < fileRecordOriginal.size()) {
+            titles += "|" + fileRecordOriginal.get(titlesIndex++).fileName;
+        }
+        OkHttpClient client = new OkHttpClient.Builder().build();
+        HttpUrl urlRelatedCategories = HttpUrl.parse(apiUrl).newBuilder()
+                .addQueryParameter("action", "query")
+                .addQueryParameter("prop", "categories")
+                .addQueryParameter("titles", titles)
+                .addQueryParameter("format", "json").build();
+        Request request = new Request.Builder().url(urlRelatedCategories).build();
+        Response response = client.newCall(request).execute();
+        JsonNode jsonNode = new ObjectMapper().readTree(response.body().string());
+        List<JsonNode> relatedCategories = new ArrayList<>();
+        List<List<String>> toCategoriesColumn = new ArrayList<>();
+        for (int i = 0; i < fileRecordOriginal.size(); i++) {
+            relatedCategories.add(jsonNode.path("query").path("pages").path(fileRecordOriginal.get(i).pageId).path("categories"));
+            List<String> categoriesPerFile = new ArrayList<>();
+            for (int j = 0; j < relatedCategories.get(i).size(); j++) {
+                categoriesPerFile.add(relatedCategories.get(i).get(j).findValue("title").asText());
+            }
+            toCategoriesColumn.add(categoriesPerFile);
+        }
+
+        return toCategoriesColumn;
+    }
+
+    /*
+     * Returns {@code true} if the iteration has more elements for
+     * which to fetch related categories.
+     * (In other words, returns {@code true} if {@link #next} would
+     * return an element rather than throwing an exception.)
+     *
+     * @return {@code true} if the iteration has more elements
+     */
+    @Override
+    public boolean hasNext() {
+
+        return iteratorFileRecords.hasNext();
+    }
+
+    /*
+     * This method iterates over each of the categories related to a file
+     * and stores them as a list in the relatedCategories parameter of
+     * each file record
+     *
+     * @return an instance of the FileRecord updated to include its related categories
+     */
+    @Override
+    public FileRecord next() {
+
+        FileRecord fileRecordNew = null;
+        filesIndex = 0;
+        // use for loop to create list of 20 .next()
+        if (filesIndex < API_TITLES_LIMIT) {
+            if (iteratorFileRecords.hasNext()) {
+                fileRecordOriginal.add(iteratorFileRecords.next());
+                filesIndex++;
+            }
+        }
+        try {
+            for (int i = 0; i < fileRecordOriginal.size(); i++) {
+                fileRecordNew = new FileRecord(fileRecordOriginal.get(i).fileName, fileRecordOriginal.get(i).pageId,
+                        getRelatedCategories(fileRecordOriginal).get(i), null);
```

Comment:
Here, you are calling `getRelatedCategories(fileRecordOriginal)` multiple times (at each loop iteration). This means that if `fileRecordOriginal` has more than one element (which it should, see above) then you will be repeating the same HTTP call many times. That is of course wasteful: the whole point of your current work is to minimize the number of HTTP calls to speed up the import.